source/extensions/filters/http/ext_proc/ext_proc.h

```diff
@@ -221,10 +246,24 @@ class FilterConfigPerRoute : public Router::RouteSpecificFilterConfig {
     return grpc_service_;
   }
 
+  const absl::optional<std::vector<std::string>>& untypedForwardingMetadataNamespaces() const {
+    return untyped_forwarding_namespaces_;
+  }
+  const absl::optional<std::vector<std::string>>& typedForwardingMetadataNamespaces() const {
+    return typed_forwarding_namespaces_;
+  }
+  const absl::optional<std::vector<std::string>>& untypedReceivingMetadataNamespaces() const {
+    return untyped_receiving_namespaces_;
+  }
+
 private:
   bool disabled_;
   absl::optional<envoy::extensions::filters::http::ext_proc::v3::ProcessingMode> processing_mode_;
   absl::optional<envoy::config::core::v3::GrpcService> grpc_service_;
+
+  absl::optional<std::vector<std::string>> untyped_forwarding_namespaces_;
```

Comment:
Please make the type inside the absl::optional and the optional iself `const`. This state is shared across threads and its const correctness is important.

Reply:
This code is following the pattern established by `processing_mode_` and `grpc_service_` above. The method `FilterConfigPerRoute::merge` is used in the `reduce` argument to [this template](https://github.com/envoyproxy/envoy/blob/main/source/common/http/utility.h#L642-L677), where a new config is constructed, merged with the const config provided.

That template was written in 2018 for the ext_authz filter and was used to implement route-level config in this filter by @gbrail in [this PR](https://github.com/envoyproxy/envoy/pull/17691), where there was some discussion about the decision to merge vs simply use most-specific. That said, it is the only place in the codebase today where the template is used, so it might merit revisiting this strategy and its dubious appearances toward thread-safety.

I think that is out of scope for this PR since the config pattern is already established. WDYT?

Reply:
So would you propose I refactor the route config merge logic to allow this field to be immutable (this would be rather large and in a separate PR) or remove initial support for route-level overrides for this feature?

I can't just slap a `const` on this and still support route-level override with the current code.

Reply:
I agree with @jbohanon here. The use of `const` here would definitely be better, but changing the merging behavior would technically represent a break, and changing the code around merging feels like a separate PR in itself to allow the original authors to weight in.

Reply:
I'm proposing this change. It would not take much time to make this fix and it will make configuration object safer. Configuration objects are immutable, so the compiler will ensure it is the case.

```suggestion
  const absl::optional<std::vector<std::string>> untyped_forwarding_namespaces_;
```

Reply:
@yanavlasov opened a separate PR for this refactor since it is out of the scope of this PR: https://github.com/envoyproxy/envoy/pull/31431